feat: pytket EncodedCircuit struct that keeps links back to the Hugr

aborgna-q opened the pull request:
Adds a `EncodedCircuit` to keep track of encoded pytket circuits with extenal links to subgraphs in the original Hugr.

Also `UnsupportedSubgraphs`....

acl-cqc commented on tket/src/serialize/pytket/extension/core.rs:
```diff
@@ -25,20 +30,50 @@ pub struct CoreDecoder;
+            ]
+            .contains(&opgroup) =>
+            {
+                let payload: UnsupportedSubgraphPayload = match serde_json::from_str(payload) {
```
Oh, no, sorry, returning Ok, right. You might do
```
let Ok(payload) = serde_json::from_str::<UnsupportedSubgraphPayload>(payload) else {
    // comment as per PR
    return Ok(DecodeStatus::Unsupported)
};
```

aborgna-q commented on tket/src/serialize/pytket/unsupported/payload.rs:
```diff
@@ -0,0 +1,195 @@
+/// Identifier for a hyper edge in the Hugr, encoded as a 64-bit hash that is
+/// independent of the original Hugr representation.
+///
+/// These are used to identify edges in the [`UnsupportedSubgraphPayload`]
```
We could technically retrieve that from the existing subgraphs when connecting two external references, but it's simpler to avoid the special case and store the extra ID.

That way we don't fail on mixed external/standalone payload types, and the structure of the Hugr is not lost if can't resolve the external reference.

aborgna-q commented on tket/src/serialize/pytket/unsupported/payload.rs:
```diff
@@ -0,0 +1,195 @@
+pub const OPGROUP_EXTERNAL_UNSUPPORTED_HUGR: &str = "EXTERNAL_UNSUPPORTED_HUGR";
+
+/// Identifier for a hyper edge in the Hugr, encoded as a 64-bit hash that is
+/// independent of the original Hugr representation.
```
rephrased

aborgna-q commented on tket/src/serialize/pytket/unsupported/payload.rs:
```diff
@@ -0,0 +1,195 @@
+pub enum UnsupportedSubgraphPayloadType {
+    /// A standalone payload, carrying the encoded HUGR subgraph.
+    Standalone {
+        /// A string envelope containing the encoded HUGR subgraph.
```
Yeah, the envelope format is self-describing so the decoding code can just `Hugr::load_str(hugr_envelope)` independently from the format used.

aborgna-q commented on tket/src/serialize/pytket/unsupported/payload.rs:
```diff
@@ -0,0 +1,195 @@
+    /// Note that this will produce incomplete hugrs when external function
+    /// calls or  non-local edges are present.
+    //
+    // TODO: Detect and deal with non-local edges? What to do there is not
```
This is an older TODO. My idea now was to include the external functions and global const defs in the hugr.
The `insert_forest` call in #1182 will look for additional global definitions and insert them as needed.

Other kind of non-local edges should give you an error, yes.

I need to implement some of that logic before making #1182 ready for review.

aborgna-q commented on tket/src/serialize/pytket/circuit.rs:
```diff
@@ -0,0 +1,307 @@
+    /// only contain an entry for this region if it was a dataflow container, or
+    /// no entries if it was not.
+    head_region: H::Node,
+    /// Circuits encoded from dataflow regions in the HUGR.
```
Yeah, _disjont_ is incorrect here as a circuit may be nested inside another's opaque subgraph.

aborgna-q commented on tket/src/serialize/pytket/unsupported.rs:
```diff
@@ -0,0 +1,151 @@
+
+    /// Returns `true` if the unsupported subgraph with the given ID exists.
+    pub fn contains(&self, id: SubgraphId) -> bool {
+        self.opaque_subgraphs.contains_key(&id)
```
Makes sense for `get` to return an option.
`contains` is a helper method that makes the API nicer, like adding `is_empty` when you have a `len` (clippy will call you out if you don't do that one :P )

aborgna-q commented:
I think I addressed all your comments, except for

- Avoiding using non-hash ids for `EncodedEdgeID`.
  https://github.com/CQCL/tket2/pull/1164#discussion_r2436301732
- Removing `IntoIter` from `&EncodedCircuit`.
  https://github.com/CQCL/tket2/pull/1164#discussion_r2436522384

Would like your input on my comments there.

acl-cqc commented on tket/src/serialize/pytket/circuit.rs:
```diff
@@ -0,0 +1,314 @@
+    ///
+    /// Auxiliary method for [`Self::from_hugr`].
+    ///
+    /// TODO: Add an option in [EncodeOptions] to run the subcircuit encoders in parallel.
```
```suggestion
    // TODO: Add an option in [EncodeOptions] to run the subcircuit encoders in parallel.
```

acl-cqc commented on tket/src/serialize/pytket/circuit.rs:
```diff
@@ -0,0 +1,314 @@
+            hugr: &impl HugrView<Node = N>,
+        ) -> Result<(), PytketEncodeError<N>> {
+            for command in commands.iter_mut() {
+                subgraphs.inline_payload(command, hugr)?;
```
`make_commands_standalone` is neat! But it seems a little odd to give *every* `Command` to the `OpaqueSubgraphs`....however, a quick look inside `inline_payload` reveals why; if only `tket_json_rs` gave us a nice enum ;-).

Perhaps rename `inline_payload` to `inline_any_payload` or `inline_if_payload`, tho?

acl-cqc commented on tket/src/serialize/pytket/circuit.rs:
```diff
@@ -0,0 +1,314 @@
+    }
+}
+
+impl<'c, 'a, H: HugrView> IntoIterator for &'c mut EncodedCircuit<'a, H> {
```
You can drop this now you have `circuits_mut` right?

acl-cqc commented on tket/src/serialize/pytket/encoder.rs:
```diff
@@ -198,21 +212,28 @@ impl<H: HugrView> PytketEncoderContext<H> {
         //
         // TODO: Test that unsupported subgraphs that don't affect any qubit/bit registers
         // are correctly encoded in pytket commands.
         while !self.unsupported.is_empty() {
```
nit: can this be `while let Some(node) = self.unsupported.iter().next()` or `....self.unsupported.first()` ?

acl-cqc commented on tket/src/serialize/pytket/encoder.rs:
```diff
@@ -198,21 +212,28 @@ impl<H: HugrView> PytketEncoderContext<H> {
             let node = self.unsupported.iter().next().unwrap();
-            let unsupported_subgraph = self.unsupported.extract_component(node);
-            self.emit_unsupported(unsupported_subgraph, circ)?;
+            let opaque_subgraphs = self.unsupported.extract_component(node);
```
This should be called, oh, `unsupported_nodes`, say
